0.0.9 (2013-12-16)
##################

*changes since 0.0.7 (2013-10-11)*

Bug Fixes
$$$$$$$$$

New Features
$$$$$$$$$$$$

Changes
$$$$$$$

Security
$$$$$$$$

Other Changes
$$$$$$$$$$$$$

* added comments to failed and broken tests [`837456ea4da3 <https://bitbucket.org/biocommons/hgvs/commits/837456ea4da3>`_]
* renamed grammars to .pymeta [`9fdcbf57ead6 <https://bitbucket.org/biocommons/hgvs/commits/9fdcbf57ead6>`_]
* consolidated g-c-p testing into a single test file; commented out putatively broken tests; DNAH11 works! [`ce8b5a899e25 <https://bitbucket.org/biocommons/hgvs/commits/ce8b5a899e25>`_]
* add forgotten sbin/fasta-seguid for commit -2 (0d29d0ea2d42) [`ca564d387db7 <https://bitbucket.org/biocommons/hgvs/commits/ca564d387db7>`_]
* fixed minor grammar bugs re: AA term and frameshift [`d25a501dca79 <https://bitbucket.org/biocommons/hgvs/commits/d25a501dca79>`_]
* added accession lookup for all of RefSeq protein [`0d29d0ea2d42 <https://bitbucket.org/biocommons/hgvs/commits/0d29d0ea2d42>`_]
* got 'make jenkins' target working [`93677d723c22 <https://bitbucket.org/biocommons/hgvs/commits/93677d723c22>`_]
* harmonized with UTA Makefile and setup.py to try to get tests working [`aa25fed0d74c <https://bitbucket.org/biocommons/hgvs/commits/aa25fed0d74c>`_]
* added biopython to setup.py [`49a6b54be6cd <https://bitbucket.org/biocommons/hgvs/commits/49a6b54be6cd>`_]
* fixed pro_eq grammar bug mentioned in `#42 <https://bitbucket.org/biocommons/hgvs/issues/42/>`_ [`8bed332027d7 <https://bitbucket.org/biocommons/hgvs/commits/8bed332027d7>`_]
* Updated DNAH11 and NEFL tests.  They run, so I'll mark as complete, but there are errors associated with the proteins [`86af61373c69 <https://bitbucket.org/biocommons/hgvs/commits/86af61373c69>`_]
* hgvsc_to_hgvsp: Fixed a delins bug [`6016c15cd3ea <https://bitbucket.org/biocommons/hgvs/commits/6016c15cd3ea>`_]
* hgvsc_to_hgvsp: Fixed bug in insertion indexing; improved exception handing [`dea365228151 <https://bitbucket.org/biocommons/hgvs/commits/dea365228151>`_]
* added misc/hgvs-shell to simplify manual testing [`b31b07896127 <https://bitbucket.org/biocommons/hgvs/commits/b31b07896127>`_]
* hgvs tests for DNAH11 and NEFL -> note protein not currently working just change if statement [`4fc4e47e1aff <https://bitbucket.org/biocommons/hgvs/commits/4fc4e47e1aff>`_]
* initial checkin for jenkins branch; want to test this in the build context [`8b937400505a <https://bitbucket.org/biocommons/hgvs/commits/8b937400505a>`_]
* Close branch c_to_p [`c78406666001 <https://bitbucket.org/biocommons/hgvs/commits/c78406666001>`_]
* Merged in c_to_p (pull request #3)
* auto merge [`95e4acd6c164 <https://bitbucket.org/biocommons/hgvs/commits/95e4acd6c164>`_]
* merged default into c_to_p [`eab0c0ff600d <https://bitbucket.org/biocommons/hgvs/commits/eab0c0ff600d>`_]
* Incorporate AASpecial; tests pass. [`e6e7595fc7c6 <https://bitbucket.org/biocommons/hgvs/commits/e6e7595fc7c6>`_]
* merge from default [`4cbd7f6fdbac <https://bitbucket.org/biocommons/hgvs/commits/4cbd7f6fdbac>`_]
* added AASpecial to handle p.=, p.?, p.0 (and parenthesized versions) [`62b892fece9a <https://bitbucket.org/biocommons/hgvs/commits/62b892fece9a>`_]
* fixed setup.py issue that caused omission of hgvs.utils on install [`8f01f7df12ef <https://bitbucket.org/biocommons/hgvs/commits/8f01f7df12ef>`_]
* Forgot to add a test file to mercurial [`8a305573421c <https://bitbucket.org/biocommons/hgvs/commits/8a305573421c>`_]
* Merged from default; fixed a test. [`52354c65da77 <https://bitbucket.org/biocommons/hgvs/commits/52354c65da77>`_]
* Make test file name more consistent [`102a2d6096b7 <https://bitbucket.org/biocommons/hgvs/commits/102a2d6096b7>`_]
* SImplified comparison in the event of a simple substitution; updated tests so the failed tests are commented out. [`1a35a8dac4d1 <https://bitbucket.org/biocommons/hgvs/commits/1a35a8dac4d1>`_]
* Reformatted Emily's test data to make it more consumer-friendly; continuous test tweaking - latest checkpoint. [`f289f13994ad <https://bitbucket.org/biocommons/hgvs/commits/f289f13994ad>`_]
* Another couple of fixes based on EH tests; checking in working version of the tests. [`1bb452a068a5 <https://bitbucket.org/biocommons/hgvs/commits/1bb452a068a5>`_]
* updated hgvsmapper with all g<->r<->c transformations [`87d9d1a880d5 <https://bitbucket.org/biocommons/hgvs/commits/87d9d1a880d5>`_]
* remove explicit class references from makeGrammar invocation, require fully-qualified class name in hgvs.ometa [`65e65931cbc0 <https://bitbucket.org/biocommons/hgvs/commits/65e65931cbc0>`_]
* close uncertainty branch [`c11fc7a522d6 <https://bitbucket.org/biocommons/hgvs/commits/c11fc7a522d6>`_]
* auto merge [`df63c09ed900 <https://bitbucket.org/biocommons/hgvs/commits/df63c09ed900>`_]
* auto merge uncertainty branch [`f82d3d8f0f83 <https://bitbucket.org/biocommons/hgvs/commits/f82d3d8f0f83>`_]
* added location uncertainty (parsing, representation, formatting, testing) [`000be417c8bc <https://bitbucket.org/biocommons/hgvs/commits/000be417c8bc>`_]
* added multifastadb code and tests [`082dac2399b9 <https://bitbucket.org/biocommons/hgvs/commits/082dac2399b9>`_]
* added chr_to_NC in utils, added c_to_g in hgvsmapper [`bf8f13f20fae <https://bitbucket.org/biocommons/hgvs/commits/bf8f13f20fae>`_]
* Name cleanup for tests [`6c2c78c34b2c <https://bitbucket.org/biocommons/hgvs/commits/6c2c78c34b2c>`_]
* Tests now play nicely with both real data and the mock data. [`f3fc11ef6c03 <https://bitbucket.org/biocommons/hgvs/commits/f3fc11ef6c03>`_]
* Add call to get_tx_seq() [`09d0b64d3251 <https://bitbucket.org/biocommons/hgvs/commits/09d0b64d3251>`_]
* Missed a rename in the tests. [`1d2e774d0656 <https://bitbucket.org/biocommons/hgvs/commits/1d2e774d0656>`_]
* Rename test classes to be a bit more consistent with their use. [`df0e226998d6 <https://bitbucket.org/biocommons/hgvs/commits/df0e226998d6>`_]
* Inserted hgvsc_to_hgvsp into hgvsmapper. [`f93887d38811 <https://bitbucket.org/biocommons/hgvs/commits/f93887d38811>`_]
* merge from default [`3727541ac989 <https://bitbucket.org/biocommons/hgvs/commits/3727541ac989>`_]
* align with developer.rst conventions on naming hgvs variants vs. strings [`c52763ee01ae <https://bitbucket.org/biocommons/hgvs/commits/c52763ee01ae>`_]
* [mq]: hgvsmapper-work [`15a5ea3e07b6 <https://bitbucket.org/biocommons/hgvs/commits/15a5ea3e07b6>`_]
* Fix tests to run in makefile context; some more documentation [`fb1bd9e76b17 <https://bitbucket.org/biocommons/hgvs/commits/fb1bd9e76b17>`_]
* revamped hgvs_c_to_p so its interface matches hgvsmapper; should make incorporation a simple matter of copying the hgvsc_to_hgvsp method in.    Updated tests accordingly.    Moved tests to top-level. [`2e0768c782c5 <https://bitbucket.org/biocommons/hgvs/commits/2e0768c782c5>`_]
* Merge from default [`2c8b4e43e4e9 <https://bitbucket.org/biocommons/hgvs/commits/2c8b4e43e4e9>`_]
* Re-arranging code for utils/staging for hgvs mapper. [`9cb4eb14e15d <https://bitbucket.org/biocommons/hgvs/commits/9cb4eb14e15d>`_]
* imported patch hgvs-utils-dir [`05923c5d4640 <https://bitbucket.org/biocommons/hgvs/commits/05923c5d4640>`_]
* Purged debug code [`3ffd4fb154c9 <https://bitbucket.org/biocommons/hgvs/commits/3ffd4fb154c9>`_]
* Ack - last checkin broke the tests; fixed accession setup [`f670c07ec0dc <https://bitbucket.org/biocommons/hgvs/commits/f670c07ec0dc>`_]
* format cleanup [`6277753a6449 <https://bitbucket.org/biocommons/hgvs/commits/6277753a6449>`_]
* Incorporate stopgap for protein accession; refactor so interface consumes data in the current UTA format; refactor tests to mimic UTA input; getting actual seq is still a placeholder. [`7913ef3851cd <https://bitbucket.org/biocommons/hgvs/commits/7913ef3851cd>`_]
* added multifastadb tool and tests [`5b9dd701fc3c <https://bitbucket.org/biocommons/hgvs/commits/5b9dd701fc3c>`_]
* merging default into c_to_p [`289769b7be43 <https://bitbucket.org/biocommons/hgvs/commits/289769b7be43>`_]
* added Rudy's AA p.= rule [`06e97542140c <https://bitbucket.org/biocommons/hgvs/commits/06e97542140c>`_]
* [mq]: grammar-relo [`6745f23488fe <https://bitbucket.org/biocommons/hgvs/commits/6745f23488fe>`_]
* added hgvs.stopgap [`48cdf47878b8 <https://bitbucket.org/biocommons/hgvs/commits/48cdf47878b8>`_]
* Close branch transcriptmapper [`9bae13dd99a9 <https://bitbucket.org/biocommons/hgvs/commits/9bae13dd99a9>`_]
* Merged in transcriptmapper (pull request #2)
* added TODO for tracking, prior to merging pull request [`a98ab0835e37 <https://bitbucket.org/biocommons/hgvs/commits/a98ab0835e37>`_]
* Basic handling of variants in non-coding regions; will return p.= in all cases; this does not handle the case where a 5'utr variant results in the creation of an upstream Met. [`4b8d2453961a <https://bitbucket.org/biocommons/hgvs/commits/4b8d2453961a>`_]
* merged with default, TM bug fixes and more tests [`ae2f8ae3b106 <https://bitbucket.org/biocommons/hgvs/commits/ae2f8ae3b106>`_]
* cleanup names (or at least make them a little more descriptive) [`f6b14466de1c <https://bitbucket.org/biocommons/hgvs/commits/f6b14466de1c>`_]
* added tm.cds_start_i in place of hard coding cds [`579438ded995 <https://bitbucket.org/biocommons/hgvs/commits/579438ded995>`_]
* refactoring [`3d4090c5b565 <https://bitbucket.org/biocommons/hgvs/commits/3d4090c5b565>`_]
* Roll back exon-specific changes and assume input is entire transcript concatenated together; retain the transcript data as recordtype [`c9216c96e812 <https://bitbucket.org/biocommons/hgvs/commits/c9216c96e812>`_]
* fix test for AA in 2nd exon [`868974db03fd <https://bitbucket.org/biocommons/hgvs/commits/868974db03fd>`_]
* Convert transcript data object to recordtype; add tests for multi-exon (in progress) [`84a00ded460e <https://bitbucket.org/biocommons/hgvs/commits/84a00ded460e>`_]
* more tests [`16cd545d0555 <https://bitbucket.org/biocommons/hgvs/commits/16cd545d0555>`_]
* additional TM fixes and more tests with multiple exons and strands [`8a770402a4b8 <https://bitbucket.org/biocommons/hgvs/commits/8a770402a4b8>`_]
* Account for transcripts w/ more than 1 exon (test input assumed one) [`2b33b7ceffc2 <https://bitbucket.org/biocommons/hgvs/commits/2b33b7ceffc2>`_]
* added some 1-exon tests [`6e80b91c0f0e <https://bitbucket.org/biocommons/hgvs/commits/6e80b91c0f0e>`_]
* Incorporate aa util and extend interval class (for test data); convert code to produce SequenceVariant objects for hgvs c to p.   Also hacked in a way to handle p.= into the grammar (should be reviewed before merge). [`b7b33f7a28af <https://bitbucket.org/biocommons/hgvs/commits/b7b33f7a28af>`_]
* bug fixes [`d526537ba5dd <https://bitbucket.org/biocommons/hgvs/commits/d526537ba5dd>`_]
* Merged default into c_to_p [`23d597f56992 <https://bitbucket.org/biocommons/hgvs/commits/23d597f56992>`_]
* added enum to transcriptmapper tests [`a7736a4e564b <https://bitbucket.org/biocommons/hgvs/commits/a7736a4e564b>`_]
* Last cleanup before merging default into here [`3be47314875b <https://bitbucket.org/biocommons/hgvs/commits/3be47314875b>`_]
* all input/output is hgvs-based. updated tests accordingly [`6c7daf36771e <https://bitbucket.org/biocommons/hgvs/commits/6c7daf36771e>`_]
* Close branch protein-variants [`bb1d20db4d9f <https://bitbucket.org/biocommons/hgvs/commits/bb1d20db4d9f>`_]
* Merged in protein-variants (pull request #1)
* hgvs.edit: fixed and improved fs handling, and added mediocre tests [`99552be8f2c8 <https://bitbucket.org/biocommons/hgvs/commits/99552be8f2c8>`_]
* hgvs.utils: added Xaa=X, Ter=*, Sec=U for aa1-to-aa3 & aa3-to-aa1 translation [`cfa29e7c4d3c <https://bitbucket.org/biocommons/hgvs/commits/cfa29e7c4d3c>`_]
* code cleaning [`c3f8b7de606c <https://bitbucket.org/biocommons/hgvs/commits/c3f8b7de606c>`_]
* finished tests for transcriptmapper [`a67c3deef07e <https://bitbucket.org/biocommons/hgvs/commits/a67c3deef07e>`_]
* finished all the g,r,c conversions adding more tests [`eccaf1c0a116 <https://bitbucket.org/biocommons/hgvs/commits/eccaf1c0a116>`_]
* More cleanup; simplify variant inserter code [`63b3d0762afa <https://bitbucket.org/biocommons/hgvs/commits/63b3d0762afa>`_]
* updated transcriptmapper to support g->r, r->g, r->c and appropriate tests [`fe1edde9744e <https://bitbucket.org/biocommons/hgvs/commits/fe1edde9744e>`_]
* merged edti-uta0 branch [`542bdb09950d <https://bitbucket.org/biocommons/hgvs/commits/542bdb09950d>`_]
* closing branch prior to merge [`1e6a974f736d <https://bitbucket.org/biocommons/hgvs/commits/1e6a974f736d>`_]
* edti: added __metaclass__ to edti.interface; added fetch_gene_info to uta0 [`eecf1f8ba945 <https://bitbucket.org/biocommons/hgvs/commits/eecf1f8ba945>`_]
* minor cleanup [`6b83ea5a28a0 <https://bitbucket.org/biocommons/hgvs/commits/6b83ea5a28a0>`_]
* variant insert tests [`fc93d04a30e0 <https://bitbucket.org/biocommons/hgvs/commits/fc93d04a30e0>`_]
* hgvs.edti: EDTI base interface and UTA0 implementation milestone [`97d0e68e1ad8 <https://bitbucket.org/biocommons/hgvs/commits/97d0e68e1ad8>`_]
* hgvs.parser: add function attributes for every rule to enable, e.g., Parser.parse_c_interval(...) [`557e6181d720 <https://bitbucket.org/biocommons/hgvs/commits/557e6181d720>`_]
* implemented p. parsing and formatting, with tests [`a7dffb4b8e56 <https://bitbucket.org/biocommons/hgvs/commits/a7dffb4b8e56>`_]
* hgvs.utils: handle case when aa string is None [`360db80cd30a <https://bitbucket.org/biocommons/hgvs/commits/360db80cd30a>`_]
* hgvs.utils: added aa_to_aa{1,3} functions to coerce to 1- or 3-letter amino acids [`8f6d9e1178b1 <https://bitbucket.org/biocommons/hgvs/commits/8f6d9e1178b1>`_]
* hgvs.utils: added protein 1-letter and 3-letter conversion [`a4bc6dead2c8 <https://bitbucket.org/biocommons/hgvs/commits/a4bc6dead2c8>`_]
* Checkpoint for new branch (hgvs c to p) [`150e1e07b7a1 <https://bitbucket.org/biocommons/hgvs/commits/150e1e07b7a1>`_]
* improved parsing of hgvs_position rules (i.e., without edits) to handle g,m,n,r,c,p types distinctly [`3af9524a1cd0 <https://bitbucket.org/biocommons/hgvs/commits/3af9524a1cd0>`_]
* added {gmn,c,r,p}_edit rule to parse variants without accesssions (e.g., c.76A>T) [`24b71f7de2bf <https://bitbucket.org/biocommons/hgvs/commits/24b71f7de2bf>`_]
* renamed DelIns class to RefAlt [`b05336c86ce2 <https://bitbucket.org/biocommons/hgvs/commits/b05336c86ce2>`_]
* renamed Variant to SequenceVariant, and instance variant seqref to ac [`44900fd9d210 <https://bitbucket.org/biocommons/hgvs/commits/44900fd9d210>`_]
* branched transcriptmapper [`f6c371dc2c25 <https://bitbucket.org/biocommons/hgvs/commits/f6c371dc2c25>`_]
* closed abandoned protein-support branch [`8263b4059042 <https://bitbucket.org/biocommons/hgvs/commits/8263b4059042>`_]
* updated parser tests to include aspirational and "reject" tests [`c2aa56c2a5b6 <https://bitbucket.org/biocommons/hgvs/commits/c2aa56c2a5b6>`_]
* [mq]: import-location-changes [`149427c0446f <https://bitbucket.org/biocommons/hgvs/commits/149427c0446f>`_]
* [mq]: import [`b9c845753cf2 <https://bitbucket.org/biocommons/hgvs/commits/b9c845753cf2>`_]
* hgvs.location: renamed location classes; added BaseOffset position for r. and c.; removed predicate methods (is_exonic, etc); [`9e44eda9d35c <https://bitbucket.org/biocommons/hgvs/commits/9e44eda9d35c>`_]
* incomplete, buggy milestone [`d675d900d194 <https://bitbucket.org/biocommons/hgvs/commits/d675d900d194>`_]
* setup.py: use full path for doc/description.rst [`577b69afd4ce <https://bitbucket.org/biocommons/hgvs/commits/577b69afd4ce>`_]
* auto merge [`99da3e7b304e <https://bitbucket.org/biocommons/hgvs/commits/99da3e7b304e>`_]
* updated CDSPosition to include datum and added tests [`e31c1cfa9f87 <https://bitbucket.org/biocommons/hgvs/commits/e31c1cfa9f87>`_]
* use get_distribution() rather than require() to fetch version [`77accc8b5609 <https://bitbucket.org/biocommons/hgvs/commits/77accc8b5609>`_]
* updated Makefile and setup.py [`bcecfde96234 <https://bitbucket.org/biocommons/hgvs/commits/bcecfde96234>`_]
* Fix for pathing to grammar.txt from within hgvs.parser.Parser [`ce591d5b8e1b <https://bitbucket.org/biocommons/hgvs/commits/ce591d5b8e1b>`_]
* modified setup.py to zipsafe false [`bec1bc319c2d <https://bitbucket.org/biocommons/hgvs/commits/bec1bc319c2d>`_]
* TODO edited online with Bitbucket
* revert directory to current after upload [`bdb6c1b24ad3 <https://bitbucket.org/biocommons/hgvs/commits/bdb6c1b24ad3>`_]
